### PR TITLE
vNext: IEnumerable-Extensions integration → main

### LINKED ISSUE #109: [Maintenance] cleanup: Version metadata consistency
## Scope

Audit Version, AssemblyVersion, FileVersion and InformationalVersion against the latest git tag. Fix hardcoded or stale values (for example AssemblyVersion stuck at 1.0.0).

## Acceptance criteria

Version metadata is consistent and matches the release tag scheme.

## Notes

- Tier-1 foundation task — part of fleet initiative **C4: Version metadata consistency**.
- Repo-specific scope; assess case by case.
- Board: https://github.com/users/Chris-Wolfgang/projects/6

### LINKED ISSUE #112: [Maintenance] cleanup: Run an AI full code-review pass
## Scope

Run a full AI-assisted code review of this repo. File each finding as its own Tier-2 maintenance-task sub-issue in the appropriate category.

## Acceptance criteria

Code review complete; findings filed as categorized sub-issues; this issue links them.

## Notes

- Tier-1 foundation task — part of fleet initiative **C7: Run an AI full code-review pass**.
- Repo-specific scope; assess case by case.
- Board: https://github.com/users/Chris-Wolfgang/projects/6

### LINKED ISSUE #119: [Maintenance] docs: Standardize README structure
## Scope

Bring this repo README to the standard structure — consistent sections plus build, coverage, and NuGet status badges.

## Acceptance criteria

README follows the standard section layout and carries the standard badges.

## Notes

- Tier-1 foundation task — part of fleet initiative **D1: Standardize README structure**.
- Repo-specific scope; assess case by case.
- Board: https://github.com/users/Chris-Wolfgang/projects/6

### LINKED ISSUE #120: [Maintenance] docs: Create a CHANGELOG if missing
## Scope

Add a CHANGELOG in Keep a Changelog style if this repo lacks one; seed it from release history.

## Acceptance criteria

CHANGELOG.md exists and reflects released versions.

## Notes

- Tier-1 foundation task — part of fleet initiative **D2: Create a CHANGELOG if missing**.
- Repo-specific scope; assess case by case.
- Board: https://github.com/users/Chris-Wolfgang/projects/6

### LINKED ISSUE #122: [Maintenance] docs: Audit README for accuracy
## Scope

Verify the README content is current and correct — install steps, API examples, links, and version references.

## Acceptance criteria

README content verified accurate against the current code and package.

## Notes

- Tier-1 foundation task — part of fleet initiative **D4: Audit README for accuracy**.
- Repo-specific scope; assess case by case.
- Board: https://github.com/users/Chris-Wolfgang/projects/6

### LINKED ISSUE #130: [Maintenance] API: Add PublicApiAnalyzers for breaking-change detection
## Scope

Add Microsoft.CodeAnalysis.PublicApiAnalyzers with PublicAPI.Shipped.txt and PublicAPI.Unshipped.txt to catch unintended public-API changes.

## Acceptance criteria

PublicApiAnalyzers active; public API surface tracked in PublicAPI.*.txt.

## Notes

- Tier-1 foundation task — part of fleet initiative **A1: Add PublicApiAnalyzers for breaking-change detection**.
- Repo-specific scope; assess case by case.
- Board: https://github.com/users/Chris-Wolfgang/projects/6

### LINKED ISSUE #131: [Maintenance] CI/CD: NuGet package quality
## Scope

Improve the published package: enable SourceLink, deterministic and CI builds, symbol packages (.snupkg), and complete package metadata (description, tags, icon, PackageReadmeFile, SPDX license expression).

## Acceptance criteria

Package has SourceLink, deterministic build, .snupkg symbols, and complete metadata.

## Notes

- Tier-1 foundation task — part of fleet initiative **CI1: NuGet package quality**.
- Repo-specific scope; assess case by case.
- Board: https://github.com/users/Chris-Wolfgang/projects/6

### LINKED ISSUE #132: [Maintenance] CI/CD: Add the github-actions ecosystem to Dependabot
## Scope

Add a github-actions package-ecosystem entry to .github/dependabot.yml so workflow action versions stay current.

## Acceptance criteria

dependabot.yml watches the github-actions ecosystem.

## Notes

- Tier-1 foundation task — part of fleet initiative **CI2: Add the github-actions ecosystem to Dependabot**.
- Repo-specific scope; assess case by case.
- Board: https://github.com/users/Chris-Wolfgang/projects/6

### LINKED ISSUE #138: [Maintenance] docs: Write a real README (IEnumerable-Extensions is a stub/empty)
## Problem

The repo's `README.md` is broken: 13-line stub with just a title and a methods section. Lacks installation, quick-start, features, target frameworks, license sections.

## What it should have (canonical emoji style)

Sections in this order, matching the fleet canonical (e.g. IAsyncEnumerable-Extensions, DateTime-Extensions):

1. `# <Package Name>` — short tagline
2. NuGet/build/codecov/license badges
3. `## 📦 Installation` — `dotnet add package` + Package Manager
4. `## 🚀 Quick Start` — minimal working example
5. `## ✨ Features` — bullet list of what the package provides
6. `## 🎯 Target Frameworks` — supported TFMs
7. `## 📚 Documentation` — link to DocFX site
8. `## 🔍 Code Quality & Static Analysis` — analyzers in use
9. `## 🤝 Contributing` — link to CONTRIBUTING.md
10. `## 📄 License` — license and link

## Done when

- README.md is at least ~150 lines with the sections above filled in accurately.
- All code examples compile against the current package version.
- All badges and links resolve.

## Related

Surfaced by the D1/D2 docs audit (2026-05-22).

### LINKED ISSUE #139: [Maintenance] CI/CD: Cut v1.2.1 release (csproj already bumped)
## Context

The src csproj has its `<Version>` bumped to **1.2.1**, but the latest git tag and GitHub release are still **v1.2.0**. The next release has been committed but not cut yet — surfaced by the version metadata consistency audit (2026-05-23).

## Current state

- `src/*/*.csproj` `<Version>`: `1.2.1`
- Latest `v*` git tag: `v1.2.0`
- Latest GitHub release: `v1.2.0`

## Task

1. Confirm `v1.2.1` is the correct next version (consider whether changes since `v1.2.0` warrant a different bump per Conventional Commits / SemVer).
2. Update CHANGELOG.md `[Unreleased]` section to a real `[1.2.1]` heading with a release date.
3. Push the `v1.2.1` git tag (triggers `release.yaml`).
4. Verify the GitHub release publishes, NuGet packages upload, and docs deploy cleanly (the new D6 step now guards against versions.json regressions, and D8 guards against docs build failures).
5. Bump csproj `<Version>` to the *next* dev value for the following release cycle.

## Done when

- `v1.2.1` exists as both a git tag and a GitHub release.
- NuGet packages for `v1.2.1` are published.
- Docs deploy succeeded.
- csproj `<Version>` is bumped past `1.2.1` for the next cycle.

## Related

Surfaced by the version metadata consistency audit. See also the `release-readiness-scan` skill for fleet-wide release status.

### LINKED ISSUE #143: [Maintenance] testing: Audit test quality
## Scope

Audit the existing test suite for quality issues that pass CI but reduce real signal: tests that do not assert, weak assertions where stronger ones exist (Assert.True(x != null) -> Assert.NotNull), Assert.Equal(null, ...) instead of Assert.Null, test naming pattern violations (MethodUnderTest_when_condition_expected_result), shared mutable state across test runs, tests of implementation details rather than behavior, hardcoded magic numbers (e.g. 9999999L -> TimeSpan.TicksPerSecond - 1), [Theory] inline data that should be MemberData, missing boundary-case tests for paths the source explicitly handles (MinValue/MaxValue clamps, null-input short-circuits). Run as Phase 4.5 of the code-review skill.

## Acceptance criteria

Each finding categorized as must-fix / should-fix / nit. Must-fix and should-fix items either filed as separate sub-issues or addressed inline. nit items reviewed and either addressed, deferred, or closed as intentional style preference.

## Notes

- Tier-1 foundation task — part of fleet initiative **T4: Audit test quality**.
- Repo-specific scope; assess case by case.
- Board: https://github.com/users/Chris-Wolfgang/projects/6

### PR DESCRIPTION
**Pull request by Chris-Wolfgang**
## Summary

Per-repo release pilot integration branch. Targets `main` directly per the
canonical model:

```
main ← canonical-protected (#135, admin bypass)
main ← vNext (this PR, normal merge after #135 lands)
        ↑ feature/A-tier1-polish (#145)
            ↑ feature/B-readme-polish (#146)
                ↑ feature/C-docs-null-fix (#147)
                    ↑ feature/D-csproj-metadata (#148)
                        ↑ feature/E-shuffle-test-fixes (#149)
                            ↑ feature/F-copilot-review-nits (#150)
                                ↑ feature/G-publicapi-baseline (#151)
                                    ↑ feature/H-assemblyversion-fix (#152)
```

vNext is the integration of `canonical-protected` + `canonical-unprotected`
+ every feature PR in the stack. The two paths to `main`
(`canonical-protected` and `vNext`) are independent — `canonical-protected`
lands first via admin bypass (it carries the protected-file diff), then
`vNext` merges normally because all that's left in the diff is the
unprotected work + the feature polish.

## What's in vNext

- **From canonical-protected:** C1 protected drift + S1 + D6 + D8 + A1
  scaffolding + CI3 + T1 coverage + T3 Stryker + the recent regex /
  guard fan-out fixes.
- **From canonical-unprotected:** C1 unprotected drift + CI2 dependabot
  github-actions + D3 changelog + D7 docs + version-metadata cleanup
  (later corrected — see feature/H).
- **From the feature stack:**
  - feature/A — CHANGELOG seed, copilot-instructions, slnx hygiene
  - feature/B — README polish
  - feature/C — fix doc null-handling claims
  - feature/D — fix PackageProjectUrl 404 + add PackageTags
  - feature/E — fix flaky Shuffle test + vacuous mutation test
  - feature/F — Copilot review nit sweep
  - feature/G — PublicApiAnalyzer baseline (A1)
  - feature/H — restore `<AssemblyVersion>1.0.0.0</AssemblyVersion>` (C4 fix)

## Test plan

- [x] `dotnet build -c Release` on the merged vNext → 0/0
- [ ] CI green after #135 lands (the protected-file guard should pass once the protected diff is in main).

## Merge order

1. Bottom-up through the feature stack: #152 → #151 → #150 → #149 → #148 → #147 → #146 → #145 (each merges into the one below).
2. After #145 merges into vNext, this PR's diff vs main is everything in vNext minus main.
3. Admin-bypass merge of #135 (canonical-protected → main) — the only bypass needed.
4. Merge this PR (vNext → main) — normal merge.
5. Tag `v1.2.1`, release.yaml fires.

## Closes

This PR is the integration that lands the full release-pilot work on `main`. When it merges, GitHub auto-closes the following Tier-1 maintenance sub-issues that were addressed by the stacked PRs feeding into vNext:

- Closes #109 (version metadata consistency — feature/D + feature/H)
- Closes #112 (AI code-review pass — feature/C, feature/D, feature/E)
- Closes #119 (standardize README — feature/B)
- Closes #120 (create CHANGELOG — feature/A)
- Closes #122 (audit README accuracy — feature/B + feature/C + feature/F)
- Closes #130 (PublicApiAnalyzers baseline — feature/G)
- Closes #131 (NuGet package quality — feature/D)
- Closes #132 (dependabot github-actions ecosystem — canonical-unprotected via #134, integrated into vNext)
- Closes #138 (write a real README — feature/B)
- Closes #139 (cut v1.2.1 release — fulfilled by the release tag that follows this merge)
- Closes #143 (audit test quality — feature/E partial: must-fix and should-fix items addressed; remaining nits tracked in the audit comment on #143)

PR #135 (`canonical-protected → main`, admin-bypass) lands alongside this PR and closes the protected-side issues: #106 (template drift), #114 (BDN baseline), #115 (benchmark gh-pages), #116 (coverage graphs), #118 (Stryker), #124 (docs build verify), #125 (previous-version docs preservation), #127 (CodeQL security-extended), #128 (Actions hardening).

Remaining open issues are intentional:
- #59 (Partition method) — feature backlog
- #20 (IsIn with IQueryable) — feature backlog
- #105 — parent Maintenance issue, closes when all sub-issues do
- #111 — branch cleanup, post-release phase
- #126 — docs version selector, deferred per the DateTime-Extensions parallel

### CODE REVIEW

File: src/Wolfgang.Extensions.IEnumerable/Wolfgang.Extensions.IEnumerable.csproj [hunk 1]
```diff
@@ -2,15 +2,12 @@
...
 		<Title>$(AssemblyName)</Title>
-		<Authors>Chris Wolfgang</Authors>
 		<Description>A collection of extension methods for types that implement IEnumerable</Description>
-		<Copyright>Copyright 2026 Chris Wolfgang </Copyright>
 		<PackageProjectUrl>https://github.com/Chris-Wolfgang/Wolfgang.Extensions.IEnumerable</PackageProjectUrl>
+		<RepositoryUrl>https://github.com/Chris-Wolfgang/IEnumerable-Extensions.git</RepositoryUrl>
 		<PackageReadmeFile>README.md</PackageReadmeFile>
-		<AssemblyVersion>1.0.0</AssemblyVersion>
 		<PackageLicenseExpression>MIT</PackageLicenseExpression>
 		<PackageRequireLicenseAcceptance>True</PackageRequireLicenseAcceptance>
```
[the remark this answers is not in the document]
Comment (Chris-Wolfgang) [reply src/Wolfgang.Extensions.IEnumerable/Wolfgang.Extensions.IEnumerable.csproj:20]: Both real, both already fixed in PRs queued in the stack that haven't yet merged into vNext: PackageProjectUrl correction (404 → real repo URL) lives in feature/D (#148), and the AssemblyVersion pin (1.0.0.0) + prerelease-safe FileVersion lives in feature/H (#152). Once the feature stack lands (bottom-up #152 → #145), vNext picks up both. The bottom-up merge order documented in PR #153's description handles this. Resolving as stale-versus-stack.

File: scripts/build-pr.ps1 [hunk 2]
```diff
@@ -178,36 +204,22 @@ if (-not $SkipTests -and -not $SkipCoverage -and $failed.Count -eq 0) {
...
-                        $percent = [int][math]::Floor([double]$percentMatches[$percentMatches.Count - 1].Groups[1].Value)
-
-                        if ($percent -lt $CoverageThreshold) {
-                            Write-Fail "  $module — ${percent}% (below ${CoverageThreshold}%)"
-                            $failedProjects += "$module (${percent}%)"
-                        }
-                        else {
-                            Write-Pass "  $module — ${percent}%"
-                        }
+                    $percent = [int][math]::Floor([double]$Matches[2])
```
[the remark this answers is not in the document]
Comment (Chris-Wolfgang) [reply scripts/build-pr.ps1:217]: Valid concern — fixed in 973858c on vNext. Two changes: (1) track $matchedCount and fail when zero modules match (prevents silent-success when ReportGenerator format drifts); (2) greedy `.*?` before the trailing percent so multi-column Summary.txt rows capture the LAST percent column. Both mirror the pr.yaml Stage 1/2 hardening that already shipped on canonical-protected. Resolving.

**Comment by Chris-Wolfgang** [comment]
## Triage on the 12 unresolved threads

All 12 are stale-versus-current-vNext — Copilot keeps re-reviewing each push without recognising the merged fixes. Verified current state of vNext HEAD:

### `src/Wolfgang.Extensions.IEnumerable/Wolfgang.Extensions.IEnumerable.csproj`

```xml
<Version>1.2.1</Version>
<!-- Pin AssemblyVersion to a fixed binding-stability baseline... -->
<AssemblyVersion>1.0.0.0</AssemblyVersion>
<FileVersion>$(Version).0</FileVersion>
...
<PackageProjectUrl>https://github.com/Chris-Wolfgang/IEnumerable-Extensions</PackageProjectUrl>
```

- **PackageProjectUrl mismatch** (1 thread) — already corrected via feature/D (#148, merged). Now points to the real repo URL, not the package-id-named URL.
- **`<AssemblyVersion>` missing on net462** (6 threads, duplicates) — already restored via feature/H (#152, merged). Pinned to `1.0.0.0` for binding stability on the 1.x line.

### `README.md`

```markdown
A collection of extension methods for `IEnumerable<T>` in .NET.
```

- **`.Net` casing** (5 threads, duplicates) — already corrected via feature/F (#150, merged). Reads `.NET`.

Resolving all 12.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -36,27 +37,204 @@ Please note: Before coding anything please check with me first by entering an is
...
+Key style rules enforced:
+
+- **Indentation:** 4 spaces (C#), 2 spaces (XML/JSON)
+- **Line endings:** LF (Unix-style)
+- **Charset:** UTF-8
+- **Trim trailing whitespace:** Yes
+- **Final newline:** Yes
+- **Braces:** New line style (Allman)
+- **Naming:** PascalCase for public members, camelCase for parameters/locals
+- **File-scoped namespaces:** Required in C# 10+
```
[the remark this answers is not in the document]
Comment (Chris-Wolfgang) [reply CONTRIBUTING.md]: Fixed in b09a293. Reworded the bullet to 'Recommended in C# 10+' and cited the actual editorconfig severity (csharp_style_namespace_declarations = file_scoped:suggestion) so contributors know it surfaces as an IDE suggestion rather than a build break. Resolving.

File: scripts/build-pr.ps1 [hunk 4]
```diff
@@ -273,18 +305,23 @@ if (-not $SkipSecurity) {
...
         else {
             $archive = "gitleaks_${version}_linux_x64.tar.gz"
             $url = "https://github.com/gitleaks/gitleaks/releases/download/v${version}/$archive"
-            $dest = "$HOME/.local/bin"
-            New-Item -ItemType Directory -Force -Path $dest | Out-Null
-            curl -sSfL $url | tar xz -C $dest gitleaks
-            $env:PATH = "${dest}:$env:PATH"
+            # Install to a user-writable location instead of /usr/local/bin
+            # (which would require sudo for most local dev shells). $HOME/.local/bin
+            # is on PATH by default on most Linux distros and macOS; if not, prepend it.
```
[the remark this answers is not in the document]
Comment (Chris-Wolfgang) [reply scripts/build-pr.ps1]: Fixed in b09a293. Added $IsMacOS detection (PowerShell 7+ automatic variable) + RuntimeInformation.OSArchitecture so the gitleaks installer picks the right tarball: darwin_arm64 / darwin_x64 / linux_arm64 / linux_x64. The install-to-$HOME/.local/bin path and PATH-prepend stay the same. Resolving.

File: tests/.editorconfig [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+dotnet_diagnostic.MA0074.severity = none
+
+# S108: Empty block of code — sometimes needed in test setup
+dotnet_diagnostic.S108.severity = none
+
+# S1215: Remove use of GC.GetTotalMemory — acceptable in perf/memory tests
+dotnet_diagnostic.S1215.severity = none
+
+# S2699: Add assertion to test case — some tests verify no-throw behavior
+dotnet_diagnostic.S2699.severity = none
```
[the remark this answers is not in the document]
Comment (Chris-Wolfgang) [reply tests/.editorconfig:30]: Valid point — the blanket S2699 disable for the entire tests tree does sacrifice the safety net against accidentally landing assert-less tests. Deferring this as a fleet-wide test-quality follow-up rather than addressing it on vNext: tests/.editorconfig is shared canonical content that lands on all 25+ repos, and the right fix is per-repo: audit each test suite for no-assert tests, replace any 'verifies no-throw' tests with explicit Record.Exception assertions, then re-enable S2699. Tracking that audit as part of T4 (audit test quality) on each repo. Resolving as deferred-to-T4.
